### PR TITLE
[New Featrue] Support Vectorization Execution Engine Interface For Doris

### PR DESCRIPTION
**Pull request by HappenLee**
1. FE vecrizied plan code
2. Function register vec function
3. Diff function nullable type
4. New thirdpart code and new thirt struct

## Proposed changes

Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Code refactor (Modify the code structure, format the code, etc...)
- [ ] Optimization. Including functional usability improvements and performance improvements.
- [ ] Dependency. Such as changes related to third-party components.
- [ ] Other.

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have created an issue on (Fix #ISSUE) and described the bug/feature there in detail
- [x] Compiling and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] If these changes need document changes, I have updated the document
- [x] Any dependent changes have been merged

### CODE REVIEW

**Comment by wangbo** [comment]
> 3. Diff function nullable type
> 4. New thirdpart code and new thirt struct
What does point 3 and point 4 specifically refer to?

File: be/src/exec/exec_node.h [hunk 1]
```diff
@@ -46,6 +45,11 @@ class TupleRow;
 class DataSink;
 class MemTracker;
 
+namespace vectorized {
+class Block;
```
Comment (wangbo) [inline be/src/exec/exec_node.h:49]: Where does class ```Block``` come from?

File: be/src/exec/schema_scanner.cpp [hunk 2]
```diff
@@ -127,7 +127,11 @@ Status SchemaScanner::create_tuple_desc(ObjectPool* pool) {
 
     for (int i = 0; i < _column_num; ++i) {
         TSlotDescriptor t_slot_desc;
-        t_slot_desc.__set_slotType(TypeDescriptor(_columns[i].type).to_thrift());
+        if (_columns[i].type == TYPE_DECIMALV2) {
+            t_slot_desc.__set_slotType(TypeDescriptor::create_decimalv2_type(27, 9).to_thrift());
```
Comment (wangbo) [inline be/src/exec/schema_scanner.cpp:131]: Why hard code ```precision``` and ```scale``` here?

File: be/src/exec/exec_node.h (hunk 1 above)
Comment (HappenLee) [reply be/src/exec/exec_node.h:49]: Just keep the interface for compile be code do not contain the real code.

File: be/src/exec/schema_scanner.cpp (hunk 2 above)
Comment (HappenLee) [reply be/src/exec/schema_scanner.cpp:131]: Now, Doris do not really use the `precision` and `scale` in BE, so here need to a TODO

File: be/CMakeLists.txt [hunk 3]
```diff
@@ -704,3 +704,8 @@ install(FILES
     ${BASE_DIR}/../conf/odbcinst.ini
     DESTINATION ${OUTPUT_DIR}/conf)
 
+
+get_property(dirs DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR} PROPERTY INCLUDE_DIRECTORIES)
```
Comment (stdpain) [inline be/CMakeLists.txt:708]: may be we should remove cmake log ?

File: be/src/exprs/aggregate_functions.cpp [hunk 4]
```diff
@@ -1925,8 +1939,7 @@ DoubleVal AggregateFunctions::knuth_var_pop_finalize(FunctionContext* ctx,
 }
 
 DecimalV2Val AggregateFunctions::decimalv2_knuth_var_pop_finalize(FunctionContext* ctx,
-                                                                  const StringVal& state_sv) {
-    DCHECK(!state_sv.is_null);
+                                                  const StringVal& state_sv) {
```
Comment (stdpain) [inline be/src/exprs/aggregate_functions.cpp:1942]: need a code format here?

File: build.sh [hunk 5]
```diff
@@ -84,7 +92,7 @@ fi
 
 eval set -- "$OPTS"
 
-PARALLEL=$[$(nproc)/4+1]
+PARALLEL=$[$(nproc)+1]
```
Comment (stdpain) [inline build.sh:95]: set PARALLEL= $[$(nproc)/4+1]

File: be/CMakeLists.txt (hunk 3 above)
Comment (HappenLee) [reply be/CMakeLists.txt:708]: yes, I will remove it

File: fe/fe-core/src/main/java/org/apache/doris/analysis/CastExpr.java [hunk 6]
```diff
@@ -17,6 +17,9 @@
 
 package org.apache.doris.analysis;
 
+import java.util.Arrays;
+import java.util.Map;
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/CastExpr.java:21]: import order

File: fe/fe-core/src/main/java/org/apache/doris/analysis/CompoundPredicate.java [hunk 7]
```diff
@@ -17,6 +17,9 @@
 
 package org.apache.doris.analysis;
 
+import com.google.common.collect.Lists;
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/CompoundPredicate.java:20]: import order

File: fe/fe-core/src/main/java/org/apache/doris/catalog/AggregateFunction.java [hunk 8]
```diff
@@ -20,6 +20,7 @@
 import static org.apache.doris.common.io.IOUtils.readOptionStringOrNull;
 import static org.apache.doris.common.io.IOUtils.writeOptionString;
 
+import avro.shaded.com.google.common.collect.ImmutableSet;
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/AggregateFunction.java:23]: wrong import

File: fe/fe-core/src/main/java/org/apache/doris/catalog/FunctionSet.java [hunk 9]
```diff
@@ -1020,12 +1121,21 @@ private void initAggregateBuiltins() {
         // count(*)
         addBuiltin(AggregateFunction.createBuiltin(FunctionSet.COUNT,
             new ArrayList<Type>(), Type.BIGINT, Type.BIGINT,
-            prefix + "9init_zeroIN9doris_udf9BigIntValEEEvPNS2_15FunctionContextEPT_",
+            prefix + "18init_zero_not_nullIN9doris_udf9BigIntValEEEvPNS2_15FunctionContextEPT_",
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/FunctionSet.java:1124]: changed?

File: fe/fe-core/src/main/java/org/apache/doris/catalog/FunctionSet.java [hunk 10]
```diff
@@ -1037,12 +1147,21 @@ private void initAggregateBuiltins() {
             // Count
             addBuiltin(AggregateFunction.createBuiltin(FunctionSet.COUNT,
                     Lists.newArrayList(t), Type.BIGINT, Type.BIGINT,
-                    prefix + "9init_zeroIN9doris_udf9BigIntValEEEvPNS2_15FunctionContextEPT_",
+                    prefix + "18init_zero_not_nullIN9doris_udf9BigIntValEEEvPNS2_15FunctionContextEPT_",
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/FunctionSet.java:1150]: changed?

File: fe/fe-core/src/main/java/org/apache/doris/catalog/FunctionSet.java [hunk 11]
```diff
@@ -1479,7 +1845,7 @@ private void initAggregateBuiltins() {
                 prefix + "13rank_finalizeEPN9doris_udf15FunctionContextERNS1_9StringValE"));
         addBuiltin(AggregateFunction.createAnalyticBuiltin( "row_number",
                 new ArrayList<Type>(), Type.BIGINT, Type.BIGINT,
-                prefix + "9init_zeroIN9doris_udf9BigIntValEEEvPNS2_15FunctionContextEPT_",
+                prefix + "18init_zero_not_nullIN9doris_udf9BigIntValEEEvPNS2_15FunctionContextEPT_",
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/FunctionSet.java:1848]: changed?

File: fe/fe-core/src/main/java/org/apache/doris/common/FeMetaVersion.java [hunk 12]
```diff
@@ -216,6 +216,8 @@
     public static final int VERSION_101 = 101;
     // add data encrypt
     public static final int VERSION_102 = 102;
+    // add vectorized to function
+    public static final int VERSION_103 = 103;
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/common/FeMetaVersion.java:220]: Not used?

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Function.java [hunk 13]
```diff
@@ -104,40 +117,49 @@
     private HdfsURI location;
     private TFunctionBinaryType binaryType;
 
+    protected NullableMode nullableMode = NullableMode.DEPEND_ON_ARGUMENT;
+
+    private boolean vectorized;
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/Function.java:122]:
```suggestion
    private boolean vectorized = false;
```

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Function.java [hunk 14]
```diff
@@ -640,6 +664,9 @@ public void readFields(DataInput input) throws IOException {
         if (hasChecksum) {
             checksum = Text.readString(input);
         }
+        if (Catalog.getCurrentCatalogJournalVersion() >= FeMetaVersion.VERSION_101) {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/Function.java:667]: remove meta version

**Review by morningman — APPROVED** [review]
LGTM

**Comment by francisoliverlee** [comment]
@HappenLee does those 2 PR all of implementation of ver-engine?

image

**Comment by HappenLee** [comment]
@francisoliverlee yes，but it only contain the some interface of vec-engine.

**Comment by hf200012** [comment]
Great, I look forward to it

**Pull request merged**